README.md
# tsl2591 sensor driver

IRNAS Zephyr compatible implementation of `tsl2591` lux driver with `i2c` support.

**IMPORTANT NOTE: for some reason first i2c transaction with sensor fails, hence we have a dummy read inside the init function that will fail. The error can be ignored.**

Interrupt support is not yet supported in the driver!
## Setup

If you do not already have them you will need to:

* [install west](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/gs_installing.html#install-west)
* [install east](https://github.com/IRNAS/irnas-east-software)

Then follow these steps:

```bash
west init -m https://github.com/IRNAS/irnas-tsl2591-driver irnas-tsl2591-driver
cd irnas-tsl2591-driver/project
west update

# set up east globally (this only needs to be done once on each machine)
east sys-setup
# install toolchain for the version of NCS used in this project
east update toolchain
```

## Building and flashing samples

To build the application samples:

```bash
cd samples/sample_name
east build -b <board_name>
```

To flash the firmware:

```bash
east flash
```

To view RTT logs:

```bash
# Run in first terminal window
east util connect

# Run in second, new terminal window
east util rtt
```

## Driver usage

This driver was written and tested for nrf-sdk v2.3.0

To install, modify your projects `west.yml` and add the following sections:

1. In `remotes`, add the following if not already added:

```yaml
 - name: irnas
   url-base: https://github.com/irnas
```

2. In the `projects` section add at the bottom (select revision you need):

```yaml
- name: irnas-tsl2591-driver 
    repo-path: irnas-tsl2591-driver
    path: irnas/irnas-tsl2591-driver
    remote: irnas
    revision: dev
```

Then run `west update` in your freshly created bash/command prompt session.

Above command will clone `irnas-tsl2591-driver` repository inside of `<project_root>/irnas`. You can now use the driver in your application projects.

In your boards DTS file or overlay file, add tsl2591 device, under selected I2C peripheral. Example of overlay implementation can be found in samples folder.

```
&i2c0 { 
	tsl2591: tsl2591@29 {
		compatible = "irnas,tsl2591";
		reg = <0x29>;
            gain = <1>;
		atime = <0>;
	};
};
```

Only reg property is required. For description of properties see the `irnas,tsl2591.yaml` file or refer to sensor configuration section.
Several senor instances are supported, provided different addresses are specified for each sensor.

User can set `gain` property, defining ADC gain with default value 1 (Medium gain (25x)):

* 0  - Low gain (1x)
* 1  - Medium gain (25x)
* 2  - High gain (428x)
* 3  - Max gain (9876x)

User can set `atime` property, defining ADC integration time with default value 0 (100 ms):

* 0  - 100 millis
* 1  - 200 millis
* 2  - 300 millis
* 3  - 400 millis
* 4  - 500 millis
* 5  - 600 millis

To enable driver functionality `CONFIG_IRNAS_TSL2591=y` must be selected. Keep in mid, it depends on `CONFIG_I2C=y`.

## Sensor API

Standard sensor api is supported. To access functionality bind to the sensor from DT definition:

```
const struct device *tsl2591_sensor = DEVICE_DT_GET(DT_NODELABEL(tsl2591));
```

then fetch new data with fetch function - this will update data for both `chanel0` and `chanel1`:

```
sensor_sample_fetch(sensor);
```

Access data with `get` function:

```
struct sensor_value val;
sensor_channel_get(sensor, SENSOR_CHAN_ALL, &val);
```

where `val1` field of `sensor_value` represents `chanel0` as `uint16_t` value and `val2` field represents `chanel1` as `uint16_t` value. For usage and conversion to LUX value, refer to datasheet.

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)

## [Unreleased]

## [0.3.0] - 2023-06-08

### Fixed

-   error reporting when measurement fails

## [0.2.0] - 2023-06-07

### Fixed

-   disable sensor at the end of the init function

## [0.1.0] - 2023-05-17

### Added

-   tsl2591 lux sensor driver
-   basic sample
-   i2c scanner

[Unreleased]: https://github.com/IRNAS/irnas-tsl2591-driver/compare/v0.3.0...HEAD

[0.3.0]: https://github.com/IRNAS/irnas-tsl2591-driver/compare/v0.2.0...v0.3.0

[0.2.0]: https://github.com/IRNAS/irnas-tsl2591-driver/compare/v0.1.0...v0.2.0

[0.1.0]: https://github.com/IRNAS/irnas-tsl2591-driver/compare/d8693b34e6b3be8241f068a3450824214940c63f...v0.1.0


samples/basic/src/main.c
/*
 * Copyright (c) 2019 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: Apache-2.0
 */

#include <stdio.h>
#include <zephyr/device.h>
#include <zephyr/devicetree.h>
#include <zephyr/drivers/sensor.h>
#include <zephyr/kernel.h>
#include <zephyr/sys/util.h>

#include <zephyr/logging/log.h>
LOG_MODULE_REGISTER(main, CONFIG_SENSOR_LOG_LEVEL);

BUILD_ASSERT(DT_NODE_EXISTS(DT_NODELABEL(tsl2591)), "tsl2591 is not defined in DT");

const struct device *tsl2591_sensor = DEVICE_DT_GET(DT_NODELABEL(tsl2591));
struct sensor_value val;

static void fetch_and_display(const struct device *sensor)
{
	if (!sensor) {
		LOG_ERR("Failed to init tsl2591!");
	}
	int err = sensor_sample_fetch(sensor);
	if (!err) {
		err = sensor_channel_get(sensor, SENSOR_CHAN_ALL, &val);
	}
	if (err < 0) {
		LOG_ERR("ERROR: tsl2591 Update failed: %d\n", err);
	} else {
		LOG_INF("ch0: %d ch1: %d", val.val1, val.val2);
	}
}

void main(void)
{
	if (!tsl2591_sensor) {
		LOG_ERR("Failed to init tsl2591!");
		return;
	} else {
		LOG_INF("Bind to tsl2591.");
	}

	if (!device_is_ready(tsl2591_sensor)) {
		LOG_ERR("Device %s is not ready", tsl2591_sensor->name);
		return;
	} else {
		LOG_INF("Device %s is ready", tsl2591_sensor->name);
	}

	while (1) {

		fetch_and_display(tsl2591_sensor);
		k_msleep(1000);
	}
}


drivers/irnas_tsl2591_driver/irnas_tsl2591.c
/** @file irnas_tsl2591.c
 *
 * @brief TSL2591 sensor driver.
 *
 * @par
 * COPYRIGHT NOTICE: (c) 2023 Irnas. All rights reserved.
 */

#define DT_DRV_COMPAT irnas_tsl2591

#include <zephyr/drivers/i2c.h>
#include <zephyr/drivers/sensor.h>
#include <zephyr/init.h>
#include <zephyr/kernel.h>
#include <zephyr/logging/log.h>
#include <zephyr/sys/__assert.h>

#include "irnas_tsl2591.h"

LOG_MODULE_REGISTER(TSL2591, CONFIG_SENSOR_LOG_LEVEL);

#define TSL2591_DEVICE_ID 0x50

static int prv_tsl2591_set_timing(const struct device *dev, enum tsl2591_adc_integration_time atime)
{
	const struct tsl2591_config *config = dev->config;

	if (!device_is_ready(config->i2c.bus)) {
		LOG_ERR("I2C bus device not ready");
		return -ENODEV;
	}

	if (i2c_reg_update_byte_dt(&config->i2c, TSL2591_COMMAND_BIT | TSL2591_REGISTER_CONFIG,
				   TSL2591_ATIME_MASK,
				   ((uint8_t)atime << TSL2591_ATIME_SHIFT)) < 0) {
		LOG_DBG("Failed to set ATIME.");
		return -EIO;
	}

	uint8_t tmp = 0x00;
	if (i2c_reg_read_byte_dt(&config->i2c, TSL2591_COMMAND_BIT | TSL2591_REGISTER_CONFIG,
				 &tmp) < 0) {
		LOG_DBG("Failed to read ATIME.");
		return -EIO;
	}
	if (((tmp & TSL2591_ATIME_MASK) >> TSL2591_ATIME_SHIFT) != atime) {
		LOG_ERR("Failed to set integration timing! Should be: %x but is: %x", atime,
			(tmp & TSL2591_ATIME_MASK) >> TSL2591_ATIME_SHIFT);
	}
	LOG_DBG("Integration time set to: %x", (tmp & TSL2591_ATIME_MASK) >> TSL2591_ATIME_SHIFT);

	return 0;
}

static int prv_tsl2591_set_gain(const struct device *dev, enum tsl2591_gain again)
{
	const struct tsl2591_config *config = dev->config;

	if (!device_is_ready(config->i2c.bus)) {
		LOG_ERR("I2C bus device not ready");
		return -ENODEV;
	}

	if (i2c_reg_update_byte_dt(&config->i2c, TSL2591_COMMAND_BIT | TSL2591_REGISTER_CONFIG,
				   TSL2591_AGAIN_MASK,
				   ((uint8_t)again) << TSL2591_AGAIN_SHIFT) < 0) {
		LOG_ERR("Failed to set AGAIN.");
		return -EIO;
	}

	uint8_t tmp = 0x00;
	if (i2c_reg_read_byte_dt(&config->i2c, TSL2591_COMMAND_BIT | TSL2591_REGISTER_CONFIG,
				 &tmp) < 0) {
		LOG_DBG("Failed to read AGAIN.");
		return -EIO;
	}
	if (((tmp & TSL2591_AGAIN_MASK) >> TSL2591_AGAIN_SHIFT) != again) {
		LOG_ERR("Failed to set gain! Should be: %x but is: %x", again,
			(tmp & TSL2591_AGAIN_MASK) >> TSL2591_AGAIN_SHIFT);
	}
	LOG_DBG("Gain set to: %x", (tmp & TSL2591_AGAIN_MASK) >> TSL2591_AGAIN_SHIFT);

	return 0;
}

static int prv_tsl2591_power_on(const struct device *dev)
{
	const struct tsl2591_config *config = dev->config;

	if (!device_is_ready(config->i2c.bus)) {
		LOG_ERR("I2C bus device not ready");
		return -ENODEV;
	}

	if (i2c_reg_update_byte_dt(&config->i2c, TSL2591_COMMAND_BIT | TSL2591_REGISTER_ENABLE,
				   TSL2591_ENABLE_POWERON, 0xff) < 0) {
		return -EIO;
	}

	return 0;
}

static int prv_tsl2591_power_off(const struct device *dev)
{
	const struct tsl2591_config *config = dev->config;

	if (!device_is_ready(config->i2c.bus)) {
		LOG_ERR("I2C bus device not ready");
		return -ENODEV;
	}

	if (i2c_reg_update_byte_dt(&config->i2c, TSL2591_COMMAND_BIT | TSL2591_REGISTER_ENABLE,
				   TSL2591_ENABLE_POWERON, 0x00) < 0) {
		return -EIO;
	}

	return 0;
}

static int prv_tsl2591_enable_als(const struct device *dev)
{
	const struct tsl2591_config *config = dev->config;

	if (!device_is_ready(config->i2c.bus)) {
		LOG_ERR("I2C bus device not ready");
		return -ENODEV;
	}

	if (i2c_reg_update_byte_dt(&config->i2c, TSL2591_COMMAND_BIT | TSL2591_REGISTER_ENABLE,
				   TSL2591_ENABLE_AEN, 0xff) < 0) {
		return -EIO;
	}

	return 0;
}

static int prv_tsl2591_dummy_transfer(const struct device *dev)
{
	const struct tsl2591_config *config = dev->config;

	if (!device_is_ready(config->i2c.bus)) {
		LOG_ERR("I2C bus device not ready");
		return -ENODEV;
	}

	/* Make test i2c transaction - for some reason first i2c transaction fails */
	struct i2c_msg msgs[1];
	uint8_t dst = 1;
	/* Send the address to read from */
	msgs[0].buf = &dst;
	msgs[0].len = 1U;
	msgs[0].flags = I2C_MSG_WRITE | I2C_MSG_STOP;

	return i2c_transfer_dt(&config->i2c, &msgs[0], 1);
}

static int tsl2591_sample_fetch(const struct device *dev, enum sensor_channel chan)
{
	struct tsl2591_data *data = dev->data;
	const struct tsl2591_config *config = dev->config;

	uint8_t ch0[2], ch1[2];

	__ASSERT_NO_MSG(chan == SENSOR_CHAN_ALL);

	/* Power on */
	prv_tsl2591_power_on(dev);

	/* Calculate minimum sleep time after power on - we need to sleep for a bit more than
	 * integration time value*/
	uint32_t sleep = 100 * config->atime + 110;
	k_sleep(K_MSEC(sleep));

	if (i2c_burst_read_dt(&config->i2c, TSL2591_COMMAND_BIT | TSL2591_REGISTER_CHAN0_LOW, ch0,
			      2)) {
		LOG_ERR("Failed to read CH0 data.");
		return -ENXIO;
	}

	data->chan0 = (ch0[1] << 8) + ch0[0];

	LOG_DBG("CH0 lsb: %x msb: %x combined: %d", ch0[0], ch0[1], data->chan0);

	if (i2c_burst_read_dt(&config->i2c, TSL2591_COMMAND_BIT | TSL2591_REGISTER_CHAN1_LOW, ch1,
			      2)) {
		LOG_ERR("Failed to read CH1 data.");
		return -ENXIO;
	}

	data->chan1 = (ch1[1] << 8) + ch1[0];

	LOG_DBG("CH1 lsb: %x msb: %x combined: %d", ch1[0], ch1[1], data->chan1);

	/* Power off */
	prv_tsl2591_power_off(dev);

	return 0;
}

static int tsl2591_channel_get(const struct device *dev, enum sensor_channel chan,
			       struct sensor_value *val)
{
	struct tsl2591_data *data = dev->data;

	__ASSERT_NO_MSG(chan == SENSOR_CHAN_ALL);

	val->val1 = data->chan0;
	val->val2 = data->chan1;

	return 0;
}

static const struct sensor_driver_api tsl2591_api = {
	.sample_fetch = &tsl2591_sample_fetch,
	.channel_get = &tsl2591_channel_get,
};

static int tsl2591_init(const struct device *dev)
{
	struct tsl2591_data *data = dev->data;
	const struct tsl2591_config *config = dev->config;

	if (!device_is_ready(config->i2c.bus)) {
		LOG_ERR("I2C bus device not ready");
		return -ENODEV;
	}

	LOG_DBG("We did find TSL2591 device with address: %x and name: %s!", config->i2c.addr,
		config->i2c.bus->name);

	data->chan0 = 0U;
	data->chan1 = 0U;

	/* Make test i2c transaction - for some reason first i2c transaction fails */
	if (prv_tsl2591_dummy_transfer(dev)) {
		LOG_DBG("Test transaction failed!");
	}

	/* Read device ID */
	uint8_t id;
	if (i2c_reg_read_byte_dt(&config->i2c, TSL2591_COMMAND_BIT | TSL2591_REGISTER_DEVICE_ID,
				 &id) < 0) {
		LOG_ERR("TSL2591 failed to read device ID.");
		return -EIO;
	}

	if (id != TSL2591_DEVICE_ID) {
		LOG_ERR("Invalid device id: %x, should be: %x", id, TSL2591_DEVICE_ID);
		return -EIO;
	}

	LOG_DBG("Read device ID: %x of TSL2591 sensor.", id);

	/* Set default timing */
	if (prv_tsl2591_set_timing(dev, config->atime)) {
		return -EIO;
	}

	/* Set default gain */
	if (prv_tsl2591_set_gain(dev, config->gain)) {
		return -EIO;
	}

	/* Enable ALS */
	if (prv_tsl2591_enable_als(dev)) {
		return -EIO;
	}

	/* Power off */
	prv_tsl2591_power_off(dev);

	return 0;
}

#define TSL2591_DEFINE(inst)                                                                       \
	static struct tsl2591_data tsl2591_data_##inst;                                            \
                                                                                                   \
	static const struct tsl2591_config tsl2591_config_##inst = {                               \
		.i2c = I2C_DT_SPEC_INST_GET(inst),                                                 \
		COND_CODE_1(DT_INST_NODE_HAS_PROP(inst, gain),                                     \
			    (.gain = DT_INST_PROP(inst, gain), ),                                  \
			    (.gain = TSL2591_GAIN_DEFAULT, ))                                      \
			COND_CODE_1(DT_INST_NODE_HAS_PROP(inst, atime),                            \
				    (.atime = DT_INST_PROP(inst, atime), ),                        \
				    (.atime = TSL2591_ATIME_DEFAULT, ))};                          \
                                                                                                   \
	SENSOR_DEVICE_DT_INST_DEFINE(inst, &tsl2591_init, NULL, &tsl2591_data_##inst,              \
				     &tsl2591_config_##inst, POST_KERNEL,                          \
				     CONFIG_SENSOR_INIT_PRIORITY, &tsl2591_api);

DT_INST_FOREACH_STATUS_OKAY(TSL2591_DEFINE)

drivers/irnas_tsl2591_driver/irnas_tsl2591.h
/** @file irnas_tsl2591.h
 *
 * @brief TSL2591 sensor driver.
 *
 * @par
 * COPYRIGHT NOTICE: (c) 2023 Irnas. All rights reserved.
 */

#ifndef IRNAS_TSL2591_H_
#define IRNAS_TSL2591_H_

#include <zephyr/device.h>
#include <zephyr/drivers/gpio.h>
#include <zephyr/drivers/i2c.h>
#include <zephyr/drivers/sensor.h>
#include <zephyr/kernel.h>

/* Command register */
#define TSL2591_COMMAND_BIT		 0xA0 // 1010 0000: bits 7 and 5 for 'command normal'
#define TSL2591_COMMAND_INT_SET		 0xE4 // Interrupt set – forces an interrupt
#define TSL2591_COMMAND_INT_CLEAR_ALS	 0xE6 // Clears ALS interrupt
#define TSL2591_COMMAND_INT_CLEAR	 0xE7 // Clears ALS and no persist ALS interrupt
#define TSL2591_COMMAND_INT_CLEAR_ALS	 0xE6 // Clears ALS interrupt
#define TSL2591_COMMAND_INT_CLEAR_ALS_NP 0xEA // Clears no persist ALS interrupt

/* Registers */
#define TSL2591_REGISTER_ENABLE		   0x00	 // Enables states and interrupts
#define TSL2591_REGISTER_CONFIG		   0x01	 // ALS gain and integration time configuration
#define TSL2591_REGISTER_THRESHOLD_AILTL   0x04	 // ALS low threshold lower byte
#define TSL2591_REGISTER_THRESHOLD_AILTH   0x05	 // ALS low threshold upper byte
#define TSL2591_REGISTER_THRESHOLD_AIHTL   0x06	 // ALS high threshold lower byte
#define TSL2591_REGISTER_THRESHOLD_AIHTH   0x07	 // ALS high threshold upper byte
#define TSL2591_REGISTER_THRESHOLD_NPAILTL 0x08	 // No Persist ALS low threshold lower byte
#define TSL2591_REGISTER_THRESHOLD_NPAILTH 0x09	 // No Persist ALS low threshold higher byte
#define TSL2591_REGISTER_THRESHOLD_NPAIHTL 0x0A	 // No Persist ALS high threshold lower byte
#define TSL2591_REGISTER_THRESHOLD_NPAIHTH 0x0B, // No Persist ALS high threshold higher byte
#define TSL2591_REGISTER_PERSIST_FILTER	   0x0C	 // Interrupt persistence filter
#define TSL2591_REGISTER_PACKAGE_PID	   0x11	 // Package Identification
#define TSL2591_REGISTER_DEVICE_ID	   0x12	 // Device Identification
#define TSL2591_REGISTER_DEVICE_STATUS	   0x13	 // Internal Status
#define TSL2591_REGISTER_CHAN0_LOW	   0x14	 // Channel 0 data, low byte
#define TSL2591_REGISTER_CHAN0_HIGH	   0x15	 // Channel 0 data, high byte
#define TSL2591_REGISTER_CHAN1_LOW	   0x16	 // Channel 1 data, low byte
#define TSL2591_REGISTER_CHAN1_HIGH	   0x17	 // Channel 1 data, high byte

/* Bit masks */
#define TSL2591_ATIME_MASK  0x07 // ADC integration time mask for TSL2591_REGISTER_CONFIG
#define TSL2591_ATIME_SHIFT 0	 // ADC integration bit shift TSL2591_REGISTER_CONFIG
#define TSL2591_AGAIN_MASK  0x30 // ALS gain mask for TSL2591_REGISTER_CONFIG
#define TSL2591_AGAIN_SHIFT 4	 // Bit shift of ALS gain

/* ENABLE mask */
#define TSL2591_ENABLE_POWERON 0x01 // Flag for ENABLE register to enable

#define TSL2591_ENABLE_AEN                                                                         \
	0x02 // ALS Enable. This field activates ALS function. Writing a one activates the ALS.
	     // Writing a zero disables the ALS.
#define TSL2591_ENABLE_AIEN                                                                        \
	0x10 // ALS Interrupt Enable. When asserted permits ALS interrupts to be generated, subject
	     // to the persist filter.
#define TSL2591_ENABLE_NPIEN                                                                       \
	0x80 // No Persist Interrupt Enable. When asserted NP Threshold conditions will generate an
	     // interrupt, bypassing the persist filter

/* LUX calculation coefficients */
#define TSL2591_LUX_DF	  (408.0F) // Lux cooefficient
#define TSL2591_LUX_COEFB (1.64F)  // CH0 coefficient
#define TSL2591_LUX_COEFC (0.59F)  // CH1 coefficient A
#define TSL2591_LUX_COEFD (0.86F)  // CH2 coefficient B

/* Enumeration for the sensor gain */
enum tsl2591_gain {
	TSL2591_GAIN_LOW = 0x00,  // low gain (1x)
	TSL2591_GAIN_MED = 0x01,  // medium gain (25x)
	TSL2591_GAIN_HIGH = 0x02, // medium gain (428x)
	TSL2591_GAIN_MAX = 0x03,  // max gain (9876x)
};

#define TSL2591_GAIN_DEFAULT TSL2591_GAIN_MED;

/* Enumeration for the sensor integration timing */
enum tsl2591_adc_integration_time {
	TSL2591_INTEGRATIONTIME_100MS = 0x00, // 100 millis
	TSL2591_INTEGRATIONTIME_200MS = 0x01, // 200 millis
	TSL2591_INTEGRATIONTIME_300MS = 0x02, // 300 millis
	TSL2591_INTEGRATIONTIME_400MS = 0x03, // 400 millis
	TSL2591_INTEGRATIONTIME_500MS = 0x04, // 500 millis
	TSL2591_INTEGRATIONTIME_600MS = 0x05, // 600 millis
};

#define TSL2591_ATIME_DEFAULT TSL2591_INTEGRATIONTIME_100MS

struct tsl2591_data {
	uint16_t chan0; // Channel 0 data IR+Visible
	uint16_t chan1; // Channel 1 data IR
};

struct tsl2591_config {
	struct i2c_dt_spec i2c;
	enum tsl2591_gain gain;
	enum tsl2591_adc_integration_time atime;
};

#endif // IRNAS_TSL2591_H_